packages/website/src/pages/docs/Internal/SqProject.mdx

```diff
@@ -0,0 +1,174 @@
+
+## APIs for reacting to project changes
+
+To be notified about `SqProject` state changes, you can use one of two approaches:
+
+1. `addEventListener`
+
+```javascript
+project.addEventListener('state', (event) => {
+    console.log(event.data);
+});
+```
+
+- `state` event will fire on any state changes
+- `action` event will fire when `SqProject` dispatches any internal action
+- `output` event will fire when `SqProject` produces a new ouptut
+
+Don't forget to clean up a listener; e.g., in a React hook:
+
+```javascript
+useEffect(() => {
+    const listener: ProjectEventListener<"action"> = (event) => {
+        console.log(event.data);
+    };
+    project.addEventListener("action", listener);
+    return () => project.removeEventListener("action", listener);
+});
+```
+
+2. `waitForOutput` helper
+
+Event listener mechanism is flexible but verbose, so we also provide this shortcut:
+
+```javascript
+const output = await project.waitForOutput(headName);
+```
+
+## Multi-module example
+
+<br />
+
+<DemoProjectStateViewer
+    headName="main"
+    headModule="mainModule"
+    modules={{
+    mainModule: `import "foo" as foo
+import "bar" as bar
+res = foo.x + bar.x`,
+    foo: `export x = 1`,
+    bar: `export y = 2`,
+    }}
+/>
+
+## Linkers
+
+`SqLinker` type is a pluggable object that describes how imports are resolved.
+
+It allows you to customize the way Squiggle modules are resolved and loaded.
+
+For example, on [Squiggle Hub](https://squigglehub.org), we use a custom linker that fetches `hub:*` modules from the server through GraphQL.
+
+In the command-line `squiggle` script, we load modules from the local filesystem.
+
+For now, [Squiggle Playground](/playground) doesn't have a linker, but in the future we plan to add a way to load modules from Squiggle Hub or other sources.
+
+## Garbage collection
+
+When you add a head to the project, it will start computing the outputs for all modules that are reachable from the head.
+
+When you update a head, the project will clean up all modules that are not reachable from the remaining set of heads.
+
+## Dependency resolving and pinning
+
+<Callout type="warn">
+    This functionality is still untested. We don't use dependency pinning on Squiggle Hub yet.
+</Callout>
+
+Imports look like this: `import "import-name" as importVar`.
+
+In this example, `"import-name"` is a string that will first be resolved by the `linker.resolve` method to the real module name.
+
+This is useful because in the command-line environment you might prefer to do relative imports: `import "./foo" as foo`, and we resolve that to the absolute path. Otherwise you'd get duplicate instances of the same module.
+
+After that, the module is loaded with `linker.loadModule` method. This module takes two parameters: the module name and its optional hash that comes from pins stored on the parent module object.
+
+If the hash is provided, the linker **must** return the module with that hash.
```

Comment:
where would the hash be provided?

Reply:
It's the second parameter on `loadModule` method. This part of documentation is for users (possibly us) who are going to implement their own custom linkers, so they shouldn't be concerned with who exactly calls `loadModule` and when (it's called by `SqProject` internals, and that's an implementation detail).

The important thing is that the hash comes from parent module's pins, but that's explained in the previous sentence.